api/v1alpha1/runner_types.go

```diff
@@ -24,8 +24,15 @@ import (
 // RunnerSpec defines the desired state of Runner
 type RunnerSpec struct {
 	// +kubebuilder:validation:MinLength=3
-	// +kubebuilder:validation:Pattern=`^[^/]+/[^/]+$`
-	Repository string `json:"repository"`
+	// +kubebuilder:validation:Pattern=`^[^/]+$`
+	Organization string `json:"organization"`
+
+	// +optional
+	// +kubebuilder:validation:Pattern=`^[^/]*$`
```

Comment:
I think this would be a breaking change. The current user has set this field to a value such as "owner/repository". However, applying this change would make such a value invalid. Can you change it so that it remains compatible?

Reply:
I totally agree. But what would be the best way then?

We could make both "optional", because you could either specify organization, repository or both. However, then there's no validation anymore on these fields - so you can make a runner without either of them and the resource will be applied (though the controller will fail).

Maybe the best option is then to allow the change, but let the controller set it to failed state if it's missing both the organization and the repository

Any thoughts?

Reply:
I agree with your thought.  I also think that if both `organization` and `repository` were not specified, the controller would need to set this resource to the failed state too.

There are four resource definitions that we should consider.

### Organization Runner Mode

With this manifest, the controller starts an organization runner.

```
spec:
  organization: summerwind
```

### Respository Runner Mode

With this manifest, the controller starts a repository runner.

```
spec:
  organization: summerwind
  repository: actions-runner-controller
```

### Compat Mode

With this manifest, the controller also starts a repository runner.

```
spec:
  repository: summerwind/actions-runner-controller
```

### Invalid resource

This is an invalid resource.

```
spec:
  organization: 
  repository: 
```

If such a resource is detected, I think it is better for the controller to set `ValidationError` to the runner state as follows. For a runner resource with a "ValidationError", the controller will not start the runner pod.

```
status:
  phase: ValidationError
  reason: "InvalidRepository"
  message: ""
```

I think that such a process should be validated using Validation Webhook, but the scope of the change would be larger, so it would be better to deal with it in a separate PR.

Reply:
The webhook is an interesting idea, but probably a bit too much for now.

We could also consider phasing out the compat mode once you move a couple versions up (deprecate it from this version on) or when you move beyond v1alpha or something

I will log a meaningful error in the meantime. Also I will update the readme with the new preferred way of using the runner spec

Reply:
It's a good point to consider a phased out. The plan I think as follows:

We will release v0.5.0 that includes this changes. At the same time, we will mark the compat mode configuration as deprecated. Then remove the compat mode support in v0.6.0 completly.

Reply:
I guess that's the cleanest solution.

In that case, I am thinking about renaming `organization` to `owner` in the RunnerSpec, because this mode has to be used for private repositories as well (and having `organization` suggests that the field is only usable for organizations - not for personal repositories)

Reply:
However, the more I think about it, the more options there are.

1. allow `organization` for org, and keep `repository` as-is for repository mode (so either allow one or the other, but _not_ both). This is very explicit about the intent of the runner and backwards compatible (no need for a phase out) - but there will be a gotcha for users that supply both
2. introduce `owner` (instead of organization) and repository (both optional, but one should be supplied). Repository will then have a different meaning (and a different format) depending on if there is an owner or not. Basically, this is how it's implemented now, except for naming (organization implies that this is only for organizations, though it's really for both)

Any maybe even more..? 

Only issue is that all scenarios that I could think of have the same downside: spec validation cannot be done, because we have two fields that depend on each other (this would require a webhook to implement fully, but that's a different story)

@summerwind I'm leaning a bit towards option 1. now (the most explicit and most similar to current implementation), but I'd like you to have the final say in the matter. Or maybe there are even better options that I haven't though of?

Reply:
Thank you for your consideration!
I would choose option 1. This is the simplest choice right now.

I don't think we should introduce `owner`. I feel `owner` is an abstract and confusing word because users choose Organization-Level or Repository-Level, and I think `owner` is an expression for implementers.

Reply:
Agreed :)

I'll implement option 1

Reply:
Ok, implementation is done, you can check if you like